File: 백준/Gold/1068. 트리/트리.cc
#include <bits/stdc++.h>
using namespace std;

int n, x, root, cnt, visited[50];
vector<int> adj[50];

void dfs(int n) {
	int tmp = 0;
	visited[n] = 1;

	for (int v : adj[n]) {
		if (!visited[v] && v != x) {
			dfs(v);
			tmp++;
		}
	}

	if (tmp == 0)
		cnt++;
}

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> n;

	for (int i = 0; i < n; i++) {
		cin >> x;
		if (x == -1)
			root = i;
		else
			adj[x].push_back(i);
	}

	cin >> x;

	if (x == root) {
		cout << 0;
		return 0;
	}

	dfs(root);

	cout << cnt;

	return 0;
}

File: 백준/Gold/1068. 트리/트리.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

		int n = Integer.parseInt(br.readLine());
		int root = -1;
		List<List<Integer>> adj = new ArrayList<>();
		for (int i = 0; i < n; i++) {
			adj.add(new ArrayList<>());
		}

		StringTokenizer st = new StringTokenizer(br.readLine());
		for (int i = 0; i < n; i++) {
			int x = Integer.parseInt(st.nextToken());
			if (x == -1) {
				root = i;
				continue;
			}

			adj.get(x).add(i);
			adj.get(i).add(x);
		}

		int rm = Integer.parseInt(br.readLine());

		if (rm == root) {
			bw.write("0");
		} else {
			Queue<Integer> q = new LinkedList<>();
			boolean[] visited = new boolean[n];
			q.add(root);
			visited[root] = true;

			int cnt = 0;
			while (!q.isEmpty()) {
				boolean flag = false;
				int x = q.poll();
				for (int nx : adj.get(x)) {
					if (visited[nx])
						continue;
					if (nx == rm)
						continue;
					flag = true;
					q.add(nx);
					visited[nx] = true;
				}

				if (!flag) {
					cnt++;
				}
			}

			bw.write(cnt + "");
		}

		bw.flush();
	}
}
